### PR TITLE
refactor: background proving

### PR DESCRIPTION
**Pull request by TalDerei**
References #324 

Currently work in progress and will require a protobuf modification to the `authorizeAndBuild` RPC method. 

The RPC methods `custodyClient.authorize` and `viewClient.witnessAndBuild` can't run simultaneously in our current construction. Instead, replace these calls with a single view service RPC method (`viewClient.authorizeAndBuild`) that handles the asynchronous tx approval. This necessitates message handling between the view server and custody server since (1) view server handles the build request, and (2) custody server handles getting the authorization. The new routing refactor supports this server-to-server message passing functionality. We'll eventually use an instance of gRPC client (`custodyClient`) in context to call `authorize` instead of our current workaround.

### CODE REVIEW

**Comment by turbocrime** [comment]
pairing yesterday, we discovered that creation of a custody service client for use in `contextValues` does not succeed. this blocks rpc init, timing out all other clients.

the failure occurs at the boundary between the `ClientConnectionManager` and `BackgroundConnectionManager`, which communicate via a `chrome.runtime.Port`. i believe this is due to a chrome runtime limitation that a script should not open a connection to itself: previously, all of our intra-extension messaging was from popup documents and such, which run independently of the background service-worker. but now we would like services running within the service-worker to talk to each other, which would violate this limitation.

currently we can just step around this by having a method impl directly call the other services' method impl when necessary, but eventually we should avoid the connection manager for impl<->impl calls.

i believe this will involve creating a `getPort` function argument for `createChannelClient` when we call it during context creation, that provides a native DOM `MessagePort`, created at the top level of the service worker. the messageport would enter the adapter directly using the same method as the background connection manager, and no larger refactor should be required

**Comment by TalDerei** [comment]
Currently, 2 popups will be trigged since the custody client `authorize` method is being called twice.

File: apps/webapp/src/state/send.ts [hunk 1]
```diff
@@ -109,10 +111,11 @@ const planWitnessBuildBroadcast = async ({ amount, recipient, selection, memo }:
...
 
+  // Remove `authorizationData` after modifying protobuf definition
   const { data: authorizationData } = await custodyClient.authorize({ plan });
   if (!authorizationData) throw new Error('no authorization data in response');
 
-  const { transaction } = await viewClient.witnessAndBuild({
+  const { transaction } = await viewClient.authorizeAndBuild({
     transactionPlan: plan,
     authorizationData,
   });
```
Comment (turbocrime) [inline apps/webapp/src/state/send.ts:121]: this is the source of the double popup - the custody service authorize impl is where the popup is requested. we call it here in send.ts, and then also in the impl

https://github.com/penumbra-zone/web/blob/0104733f44e08d1d27d1e9c7a110da5ddcf870b1/packages/router/src/grpc/view-protocol-server/authorize-and-build.ts#L13-L17

we could simply remove the authorize call here in send.ts, and pass an empty authorizationdata to authorizeandbuild.

Comment (turbocrime) [reply apps/webapp/src/state/send.ts:121]: cc penumbra-zone/penumbra#3618

Comment (TalDerei) [reply apps/webapp/src/state/send.ts:121]: I recalled protobuf fields are optional by default, and removing the `AuthorizeData` field entirely seemed to work.

**Comment by TalDerei** [comment]
Web implementation will consume the new streaming interfaces per https://github.com/penumbra-zone/web/issues/384 and https://github.com/penumbra-zone/penumbra/pull/3559 separately to limit scope.

**Review by turbocrime — COMMENTED** [review]
needs a bit more cleanup but the working details are correct now

File: apps/extension/src/clients/extension-page.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    {
+      type: 'INIT_CHANNEL_CLIENT' as typeof InitChannelClientDataType,
+      port: transferPort,
+    } as InitChannelClientMessage,
+    [transferPort],
+  );
+  return port;
+};
+
+export const createExtensionPageClient = <S extends ServiceType>(serviceType: S) =>
```
[the remark this answers is not in the document]
Comment (TalDerei) [reply apps/extension/src/clients/extension-page.ts:34]: `ExtensionPage` refers to a webpage (for instance a content script or popup page) associated with the extension. `createExtensionPageClient` inits a client object for establishing communication with the extension's services.

[the remark this answers is not in the document]
Comment (grod220) [reply apps/extension/src/clients/extension-page.ts:34]: Please add code comment above 🙏

File: apps/extension/src/clients/extension-service.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+ * router entry, the returned function satisfies the `getPort` parameter
+ * required by `createChannelTransport`.
+ *
+ * TODO: this could probably be provided by ClientConnectionManager
+ *
+ * @param entry router entry function
+ * @returns getPort parameter for creating a channel client
+ */
+export const directEntry = (entry: ChromeRuntimeHandlerFn) => {
+  const getDirectEntryPort = (): MessagePort => {
```
[the remark this answers is not in the document]
Comment (TalDerei) [reply apps/extension/src/clients/extension-service.ts:26]: simplified

File: apps/extension/tsconfig.json [hunk 4]
```diff
@@ -4,6 +4,6 @@
   "compilerOptions": {
     "outDir": "./dist",
     "declaration": false,
-    "declarationMap": false
-  }
+    "declarationMap": false,
+  },
```
[the remark this answers is not in the document]
Comment (TalDerei) [reply apps/extension/tsconfig.json:8]: these were added by prettier during formatting

File: apps/webapp/src/state/send.ts [hunk 5]
```diff
@@ -85,7 +85,9 @@ export const createSendSlice = (): SliceCreator<SendSlice> => (set, get) => {
   };
 };
 
+// rename to PlanAuthorizeWitnessBuildBroadcast or AuthorizeWitnessBuild?
 const planWitnessBuildBroadcast = async ({ amount, recipient, selection, memo }: SendSlice) => {
+  console.log('Entered planWitnessBuildBroadcast!');
```
[the remark this answers is not in the document]
Comment (TalDerei) [reply apps/webapp/src/state/send.ts:90]: removed

File: apps/webapp/src/state/send.ts [hunk 6]
```diff
@@ -85,7 +85,9 @@ export const createSendSlice = (): SliceCreator<SendSlice> => (set, get) => {
   };
 };
 
+// rename to PlanAuthorizeWitnessBuildBroadcast or AuthorizeWitnessBuild?
 const planWitnessBuildBroadcast = async ({ amount, recipient, selection, memo }: SendSlice) => {
```
[the remark this answers is not in the document]
Comment (TalDerei) [reply apps/webapp/src/state/send.ts:89]: changed to use generic helper

File: packages/router/src/grpc/view-protocol-server/authorize-and-build.ts
[the remark this answers is not in the document]
Comment (TalDerei) [reply packages/router/src/grpc/view-protocol-server/authorize-and-build.ts:1]: I think splitting up the functionality is cleaner and more idiomatic. I implemented the suggest changes.

File: apps/extension/tsconfig.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (TalDerei) [reply apps/extension/tsconfig.json:8]: I resolved the trailing commas

File: apps/extension/src/clients/extension-service.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  createChannelTransport,
+  isTransportMessage,
+} from '@penumbra-zone/transport';
+import { ChromeRuntimeHandlerFn } from '@penumbra-zone/transport/src/chrome-runtime/adapter';
+import { typeRegistry } from '@penumbra-zone/types/src/registry';
+import { transportOptions } from './transport-options';
+
+/**
+ * This enables a channelTransport to communicate directly with a router running
+ * in the same script, without going through the connection manager. Given
```
[the remark this answers is not in the document]
Comment (grod220) [reply apps/extension/src/clients/extension-service.ts:16]: Can you clarify in documentation "same script"?

File: apps/webapp/src/state/helper.ts
Comment (grod220) [inline apps/webapp/src/state/helper.ts:1]: nit: can you rename to `helpers.ts`?

**Review by grod220 — APPROVED** [review]
Can you rebase to the lastest main? Great job!

File: apps/webapp/src/state/helper.ts
Comment (TalDerei) [reply apps/webapp/src/state/helper.ts:1]: changed and rebased from main

**Pull request merged**